File: Level 3/단어 변환.swift
import Foundation

// 각 단어는 알파벳 소문자로만 이루어져 있고, 길이가 동일
func isChanged(_ now:String, _ word:String) -> Bool {
    var count : Int = 0
    
    if now.count != word.count {
        return false
    }
    
    // 문자열을 배열로 변환
    let nowArr = Array(now)
    let wordArr = Array(word)
    for i in 0..<nowArr.count {
        if nowArr[i] != wordArr[i] {
            count += 1
        }
        
        if count > 1 {
            return false
        }
    }
    
    // 다른 알파벳이 1개이면 변환 가능
    return true
}

func solution(_ begin:String, _ target:String, _ words:[String]) -> Int {
    var queue : [String] = [] // 단어를 담을 큐
    queue.append(begin)
    
    var isVisited = Array(repeating: false, count: words.count) // 방문 정보
    var answer : Int = 0 // 변환 결과
    
    if !words.contains(target) { // words 안에 target이 존재하지 않으면 변환 불가
        return 0
    }
    
    while !queue.isEmpty {
        let now = queue.popLast()!
        if now == target { // 변환이 완료되면 while문 종료
            break
        }
        
        answer += 1
        for i in 0..<words.count {
            if !isVisited[i] && isChanged(now, words[i]) {
                isVisited[i] = true // 방문 처리
                queue.append(words[i])
            }
        }
    }
    
    return answer
}

File: Level 3/단어 변환.java
import java.util.PriorityQueue;
import java.util.Arrays;
import java.util.Queue;
import java.util.LinkedList;

class Solution {
    public boolean isChanged(String str, String word) {
        int count = 0;
        int[] strArr = new int[str.length()];
        int[] wordArr = new int[word.length()];
        
        // 문자열을 한 글자씩 배열의 원소로 저장
        for (int i = 0; i < str.length(); i++) {
            if (str.charAt(i) != word.charAt(i)) { // 글자가 다른 경우 count 증가
                count++;
            }
        }
        
        return count == 1 ? true : false;
    }
    
    // BFS
    public int solution(String begin, String target, String[] words) {
        Queue<Node> queue = new LinkedList<>();
        queue.offer(new Node(begin, 0)); // 변환을 시작할 단어를 큐에 넣기, 아직 0단계
        
        boolean[] isVisited = new boolean[words.length];
        
        while(!queue.isEmpty()) {
            Node node = queue.poll();
            
            if (node.str.equals(target)) return node.count;
            
            for (int i = 0; i < words.length; i++) {
                if (isVisited[i]) continue;
                
                if (isChanged(node.str, words[i])) {
                    queue.offer(new Node(words[i], node.count+1)); // words[i] 변환까지 몇 단계를 거쳤는지 1을 더해 노드에 저장
                    isVisited[i] = true; // 방문 처리
                }
            }
        }
        
        return 0; // target이 words 배열에 포함되어있지 않은 경우
    }
    
    public static class Node {
        String str;
        int count;
        
        Node(String str, int count) {
            this.str = str;
            this.count = count;
        }
    }
}
